src/natcap/invest/habitat_quality.py

```diff
@@ -551,264 +620,337 @@ def execute(args):
-            ValueError if a valid raster file could not be found.
+    Args:
+        raster_path_band (tuple): a 2 tuple of the form
+            (filepath to raster, band index).
 
-    Return:
-        the full path, plus extension, to the valid raster.
+    Returns:
+        None
     """
-    # Turning on exceptions so that if an error occurs when trying to open a
-    # file path we can catch it and handle it properly
-
-    def _error_handler(*_):
-        """A dummy error handler that raises a ValueError."""
-        raise ValueError()
-    gdal.PushErrorHandler(_error_handler)
-
-    # a list of possible extensions for raster datasets. We currently can
-    # handle .tif and directory paths
-    possible_ext = ['', '.tif', '.img']
-
-    # initialize dataset to None in the case that all paths do not exist
-    dataset = None
-    for ext in possible_ext:
-        full_path = path + ext
-        if not os.path.exists(full_path):
-            continue
-        try:
-            dataset = gdal.OpenEx(full_path, gdal.GA_ReadOnly)
-            break
-        except ValueError:
-            # If GDAL can't open the raster, our GDAL error handler will be
-            # executed and ValueError raised.
-            continue
+    raster_path = raster_path_band[0]
+    # declare a set to store unique codes from raster
+    raster_unique_lucodes = set()
 
-    gdal.PopErrorHandler()
+    for _, raster_block in pygeoprocessing.iterblocks(raster_path_band):
+        raster_unique_lucodes.update(numpy.unique(raster_block))
 
-    # If a dataset comes back None, then it could not be found / opened and we
-    # should fail gracefully
-    if dataset is None:
-        if raise_error:
-            raise ValueError(
-                'There was an Error locating a threat raster in the input '
-                'folder. One of the threat names in the CSV table does not '
-                'match to a threat raster in the input folder. Please check '
-                'that the names correspond. The threat raster that could not '
-                'be found is: %s' % path)
-        else:
-            full_path = None
+    # Remove the nodata value from the set of landuser codes.
+    nodata = pygeoprocessing.get_raster_info(raster_path)['nodata'][0]
+    raster_unique_lucodes.discard(nodata)
 
-    dataset = None
-    return full_path
+    return raster_unique_lucodes
 
 
-def raster_pixel_count(raster_path):
+def _raster_pixel_count(raster_path_band):
     """Count unique pixel values in raster.
 
-    Parameters:
-        raster_path (string): path to a raster
+    Args:
+        raster_path_band (tuple): a 2 tuple of the form
+            (filepath to raster, band index).
 
     Returns:
         dict of pixel values to frequency.
     """
-    nodata = pygeoprocessing.get_raster_info(raster_path)['nodata'][0]
+    nodata = pygeoprocessing.get_raster_info(
+                raster_path_band[0])['nodata'][0]
```

Comment:
I see below for `iterblocks` you're using `raster_path_band` as given, but here it assumes the nodata is whatever's in the first band. To be consistent, can you change this line to `raster_path_band[0])['nodata'][raster_path_band[1]]`?

Reply:
Hey Rich, I think there's a discrepancy in knowing a band number and assuming that band number matches the returned sequence index of `get_raster_info(path)['nodata']`. For instance in this case the band number will be `1`, but `pgp.get_raster_info(raster_path_band[0])['nodata'][1]` will error because that result is 0 indexed. The more I think about it, the only way to really match `nodata` with `band number` is to get a list of the band numbers sorted.

That seems overkill here, and I will update the docstring to indicate this should be a single band raster input:

Reply:
Oop sorry, I meant `raster_path_band[0])['nodata'][raster_path_band[1]-1]`! `raster_path_band[1]` is the band ID, but the result is a list starting from band one indexed at 0.

Reply:
> @richpsharp I think I addressed all your change requests and comments. I had a really good pun for what to write here, like: Thanks for teeing these up Rich, I knocked em all out. But it was better than that, but I forgot it.

Hi Doug, thanks for doing all that. I still have one issue on that band ID though. I'm worried I'm not understanding what the issue is, so if it doesnt' make sense let's chat. Otherwise, we use that `...raster_path_band[0])['nodata'][raster_path_band[1]-1]` paradigm all the time.